pkg/controller/provider-network.go

```diff
@@ -4,126 +4,58 @@ import (
-		if k8serrors.IsNotFound(err) {
-			return nil
-		}
-		return err
-	}
-
 	nodes, err := c.nodesLister.List(labels.Everything())
 	if err != nil {
 		klog.Errorf("failed to list nodes: %v", err)
-		return err
-	}
-
-	if providerNetworkIsReady(pn, nodes) != pn.Status.Ready {
-		newPn := pn.DeepCopy()
-		newPn.Status.Ready = !pn.Status.Ready
-		_, err = c.config.KubeOvnClient.KubeovnV1().ProviderNetworks().UpdateStatus(context.Background(), newPn, metav1.UpdateOptions{})
-		if err != nil {
-			klog.Errorf("failed to update status of provider network %s: %v", pn.Name, err)
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (c *Controller) resyncProviderNetworkStatus() {
-	nodeList, err := c.nodesLister.List(labels.Everything())
-	if err != nil {
-		klog.Errorf("failed to get nodes %v", err)
-		return
-	}
-	pnList, err := c.providerNetworksLister.List(labels.Everything())
-	if err != nil {
-		klog.Errorf("failed to get provider networks %v", err)
 		return
 	}
 
-	for _, pn := range pnList {
-		if providerNetworkIsReady(pn, nodeList) != pn.Status.Ready {
-			newPn := pn.DeepCopy()
-			newPn.Status.Ready = !newPn.Status.Ready
-			_, err = c.config.KubeOvnClient.KubeovnV1().ProviderNetworks().UpdateStatus(context.Background(), newPn, metav1.UpdateOptions{})
-			if err != nil {
-				klog.Errorf("failed to update status of provider network %s: %v", pn.Name, err)
+	for _, pn := range pns {
+		var readyNodes, notReadyNodes []string
+		for _, node := range nodes {
+			if util.ContainsString(pn.Spec.ExcludeNodes, node.Name) {
+				pn.Status.RemoveNodeConditions(node.Name)
+				continue
+			}
+			if node.Labels[fmt.Sprintf(util.ProviderNetworkReadyTemplate, pn.Name)] == "true" {
+				pn.Status.SetNodeReady(node.Name, "InitOVSBridgeSucceeded", "")
+				readyNodes = append(readyNodes, node.Name)
+			} else {
+				pods, err := c.config.KubeClient.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
+					LabelSelector: "app=kube-ovn-cni",
+					FieldSelector: fmt.Sprintf("spec.nodeName=%s", node.Name),
+				})
+				if err != nil {
+					klog.Errorf("failed to list pod: %v", err)
+				}
+
+				var errMsg string
+				if len(pods.Items) == 1 && pods.Items[0].Annotations != nil {
+					errMsg = pods.Items[0].Annotations[fmt.Sprintf(util.ProviderNetworkErrMessageTemplate, pn.Name)]
+				}
+				pn.Status.SetNodeNotReady(node.Name, "InitOVSBridgeFailed", errMsg)
+				notReadyNodes = append(notReadyNodes, node.Name)
 			}
 		}
-	}
-}
-
-func providerNetworkIsReady(pn *kubeovnv1.ProviderNetwork, nodes []*corev1.Node) bool {
-	for _, node := range nodes {
-		if !util.ContainsString(pn.Spec.ExcludeNodes, node.Name) &&
-			!util.ContainsString(pn.Status.ReadyNodes, node.Name) {
-			return false
+		pn.Status.ReadyNodes = readyNodes
+		pn.Status.NotReadyNodes = notReadyNodes
+		pn.Status.Ready = (len(notReadyNodes) == 0)
+		if _, err = c.config.KubeOvnClient.KubeovnV1().ProviderNetworks().UpdateStatus(context.Background(), pn, metav1.UpdateOptions{}); err != nil {
```

Comment:
Before applying update, we can check whether the status has updates. If no changes occurs, we can skip it.